src/dsymbol/modulecache.d

```diff
@@ -436,10 +445,92 @@ private:
 
 /// Wrapper to check some attribute of a path, ignoring errors
 /// (such as on a broken symlink).
-private static bool existsAnd(alias fun)(string fn)
+private static bool existsAnd(alias fun)(string file)
 {
 	try
-		return fun(fn);
+		return fun(file);
 	catch (FileException e)
 		return false;
 }
+
+/// same as getAttributes without throwing
+/// Returns: true if exists, false otherwise
+private static bool getFileAttributesFast(R)(R name, uint* attributes)
+{
+	version (Windows)
+	{
+		import std.internal.cstring : tempCStringW;
+		import core.sys.windows.winnt : INVALID_FILE_ATTRIBUTES,
+			FILE_ATTRIBUTE_DIRECTORY;
+		import core.sys.windows.winbase : GetFileAttributesW;
+
+		auto namez = tempCStringW(name);
+		static auto trustedGetFileAttributesW(const(wchar)* namez) @trusted
+		{
+			return GetFileAttributesW(namez);
+		}
+		*attributes = trustedGetFileAttributesW(namez);
+		return *attributes != INVALID_FILE_ATTRIBUTES;
+	}
+	else version (Posix)
+	{
+		import core.sys.posix.sys.stat : stat, stat_t, S_IFMT, S_IFREG, S_IFDIR;
+		import std.internal.cstring : tempCString;
+
+		auto namez = tempCString(name);
+		static auto trustedStat(const(char)* namez, out stat_t statbuf) @trusted
+		{
+			return stat(namez, &statbuf);
```

Comment:
You may want to use lstat instead because it returns information about the symlink itself, not what it is refers to.

Reply:
we should support symlinks as files/folders depending on what they point to, stat is the proper method to do that (no special handling for symlinks, act like they are normal files)

This is also consistent with the previous behavior when using `isDir`, `isFile` which uses a normal stat.

Reply:
I am sorry, I should have been clearer in what I meant. The implementation of isSymlink is the problem, not the other two. Phobos for example uses getLinkAttributes in its isSymlink implementation.

```d
    import core.sys.posix.sys.stat : stat, lstat, stat_t;
    import std.string : toStringz;
    import std.file : remove, symlink, mkdir;
    import std.stdio : File, writefln;

    static void info(string name) {
        stat_t statAttr;
        stat_t lstatAttr;

        assert(stat(name.toStringz, &statAttr) == 0);
        assert(lstat(name.toStringz, &lstatAttr) == 0);
        writefln("%s %x %s %X %s", name, statAttr.st_mode, attrIsSymlink(statAttr.st_mode), lstatAttr.st_mode, attrIsSymlink(lstatAttr.st_mode));
    }

    const fname = "test_file";
    const dname = "test_dir";
    const sname = "test_symlink";

    File(fname, "w").write("foo");
    mkdir(dname);

    symlink(fname, sname);
    info(sname);

    remove(sname);
    symlink(dname, sname);
    info(sname);

    symlink(sname, sname ~ "_symlink2");
    info(sname ~ "_symlink2");
```
output:
```sh
test_symlink 81b4 false A1FF true
test_symlink 41fd false A1FF true
test_symlink_symlink2 41fd false A1FF true
```

As you can see stat can't be used to detect if the file itself is a symlink, only lstat can do that.

Reply:
we don't use isSymlink, this is only used for isFile and isDir in this code (and it's private)

if we have code like
```
mkdir source
touch source/foo.d
ln -s source somewhere
```
then whe want "somewhere" to behave like a folder and "somewhere/foo.d" to behave like a file. We don't want any special handling for symlinks in the import resolver code. The return value from the `stat` function is exactly what we want here.